recirq/otoc/parallel_xeb.py

```diff
@@ -0,0 +1,725 @@
+            correponding to a list of sets of qubits with the same length as
+            max(cycles). For each cycle, single-qubit gates outside the first
+            lightcone are either removed or replaced with a spin-echo pulse.
+            Single-qubit gates outside the second lightcone, if specified,
+            are always removed.
+        echo_indices: An array with the same dimension as sq_rand_nums and
+            random integer values of 1, 2, 3 or 4. They specify the spin-echo
+            pulses applied to qubits outside the first lightcone, which can
+            be +/-X or +/-Y gates.
+
+    Returns:
+        A list of random circuits, each containing a specified number of cycles.
+    """
+    if light_cones is not None:
+        if len(light_cones) > 2:
+            raise ValueError('light_cones may only have length 1 or 2')
+
+    if benchmark_ops is not None:
+        num_d = len(benchmark_ops)
+    else:
+        num_d = 0
+    max_cycles = max(cycles)
+
+    single_rots, indices = _random_rotations(
+        qubits, max_cycles, random_seed, sq_rand_nums, light_cones,
+        echo_indices, z_rotations_only=z_only)
+
+    all_circuits = []  # type: List[cirq.Circuit]
+    for num_cycles in cycles:
+        circuit_exp = cirq.Circuit()
+        for i in range(num_cycles):
+            c = i + 1 if not reverse else num_cycles - i
+            if ancilla is not None and cycles_per_echo is not None:
+                if c % cycles_per_echo == 0:
+                    op_list = [cirq.Y(ancilla)]
+                    op_list.extend(single_rots[i])
+                    rots = cirq.Moment(op_list)
+                else:
+                    rots = cirq.Moment(single_rots[i])
+            else:
+                rots = cirq.Moment(single_rots[i])
+            if reverse:
+                if benchmark_ops is not None:
+                    circuit_exp.append(benchmark_ops[i % num_d])
+                circuit_exp.append(rots, strategy=cirq.InsertStrategy.NEW)
+            else:
+                circuit_exp.append(rots, strategy=cirq.InsertStrategy.NEW)
+                if benchmark_ops is not None:
+                    circuit_exp.append(benchmark_ops[i % num_d])
+        all_circuits.append(circuit_exp)
+    return all_circuits, indices
+
+
+def parallel_xeb_fidelities(
+        all_qubits: List[Tuple[int, int]],
+        num_cycle_range: Sequence[int],
+        measured_bits: List[List[List[np.ndarray]]],
+        scrambling_gates: List[List[np.ndarray]],
+        fsim_angles: Dict[str, float],
+        interaction_sequence: Optional[List[
+            Set[Tuple[Tuple[float, float], Tuple[float, float]]]]] = None,
+        gate_to_fit: str = 'iswap',
+        num_restarts: int = 3,
+        num_points: int = 8,
+        plot_individual_traces: bool = False,
+        plot_histograms: bool = False,
+        save_directory: Optional[str] = None
+) -> Tuple[Dict[Tuple[Tuple[int, int], Tuple[int, int]], cirq.Circuit],
```

Comment:
Make a dataclass to encapsulate the return value. This is too complex otherwise

Reply:
Done